presto-spi/src/main/java/com/facebook/presto/spi/block/RowBlockBuilder.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.spi.block;
+
+import com.facebook.presto.spi.type.Type;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+import static com.facebook.presto.spi.block.BlockUtil.calculateBlockResetSize;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class RowBlockBuilder
+        extends AbstractRowBlock
+        implements BlockBuilder
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(RowBlockBuilder.class).instanceSize();
+
+    @Nullable
+    private BlockBuilderStatus blockBuilderStatus;
+
+    private int positionCount;
+    private int[] columnBlockOffsets;
```

Comment:
Since it's the offset in field blocks, it kind of contain the information that the offset is for entries 

i.e. position-based offset in field block is the entry-based offset in parent block :)